R/animint.R

```diff
@@ -22,11 +22,54 @@ parsePlot <- function(meta){
       plot.meta$scales[[sc$aesthetics]] <- makeScale(sc)
     }
   }
+  
+  ## Export axis specification as a combination of breaks and
+  ## labels, on the relevant axis scale (i.e. so that it can
+  ## be passed into d3 on the x axis scale instead of on the
+  ## grid 0-1 scale). This allows transformations to be used
+  ## out of the box, with no additional d3 coding.
+  theme.pars <- ggplot2:::plot_theme(meta$plot)
+  
+  ## No legend if theme(legend.postion="none").
+  plot.meta$legend <- if(theme.pars$legend.position != "none"){
+    getLegendList(meta$built)
+  }
+  
+  ## save each layer
   for(layer.i in seq_along(meta$plot$layers)){
     ##cat(sprintf("%4d / %4d layers\n", layer.i, length(meta$plot$layers)))
 
     ## This is the layer from the original ggplot object.
     L <- meta$plot$layers[[layer.i]]
+    
+    ## If any legends are specified, add showSelected aesthetic
+    for(legend.i in seq_along(plot.meta$legend)) {
+      # the type of legend: colour, fill, etc
+      legend_type <- plot.meta$legend[[legend.i]]$legend_type
+      # the name of the variable used in this legend
+      var_name <- plot.meta$legend[[legend.i]]$vars
+      # the actual values for that variable
+      var <- L$data[[var_name]]
+      # checking if it is a discrete variable
+      if(plyr::is.discrete(var)) {
+        # adding a showSelected aesthetic for each legend entry
+        for(i in legend_type) {
+          temp_name <- paste0("showSelected", i)
+          L$mapping[[temp_name]] <- as.symbol(var_name)
+#           browser()
+        }
+        # if first is not specified, add all to first
+        if(is.null(meta$first[[var_name]])) {
+          meta$first[[var_name]] <- unique(var)
+        }
+        # if selectortypes is not specified, set it to multiple
+        if(is.null(meta$selector.types[[var_name]])) {
+          meta$selector.types[[var_name]] <- "multiple"
+        }
```

Comment:
the test that is currently failing (test-panels.R, "renderer can handle no grid lines" `Error: length(first) == 1 is not TRUE`) is because the sex_smoker selector was inferred to be single selection before this. 

Careful: `meta$selector.types$sex_smoker` is the user-specified value (undefined in at the beginning of this test but set to "multiple" by your code), and `meta$selectors$sex_smoker$type` is the inferred type which is actually saved to plot.json and used by the renderer.